File: Java/src/covariance/maths/CovGraph.java
package covariance.maths;

import covariance.abs.AbsEdge;
import covariance.abs.AbsGraph;
import covariance.abs.AbsVertex;


public class CovGraph extends AbsGraph
{
    public CovGraph(String name)
    {
        super("Graph " + name);
    }

    public void addVertex(CovVertex v)
    {
        super.addVertex(v);        
    }

    public void addEdge(CovEdge e, CovVertex v1, CovVertex v2)
    {
        super.addEdge(e, v1, v2);        
    }
    
    @Override
    public void addVertex(AbsVertex v)
    {
        try
        {
            addVertex((CovVertex) v);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public void addEdge(AbsEdge a, AbsVertex v1, AbsVertex v2)
    {
        try
        {
            addEdge((CovEdge) a, (CovVertex) v1, (CovVertex) v2);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }
}

File: scala/src/covariance/maths/CovGraph.scala
package covariance.maths

import covariance.abs.AbsEdge
import covariance.abs.AbsGraph
import covariance.abs.AbsVertex

class CovGraph(name: String) extends AbsGraph("Graph " + name) {

  def addVertex(v: CovVertex): Unit = {
    super.addVertex(v)
  }

  def addEdge(e: CovEdge, v1: CovVertex, v2: CovVertex): Unit = {
    super.addEdge(e, v1, v2)
  }

  override def addVertex(v: AbsVertex): Unit = {
    try {
      addVertex(v.asInstanceOf[CovVertex])
    } catch {
      case e: Exception =>
        e.printStackTrace()

    }
  }

  override def addEdge(a: AbsEdge, v1: AbsVertex, v2: AbsVertex): Unit = {
    try {
      addEdge(a.asInstanceOf[CovEdge], v1.asInstanceOf[CovVertex], v2.asInstanceOf[CovVertex])
    } catch {
      case e: Exception =>
        e.printStackTrace()

    }
  }

}
